File: rust/src/gender.rs
//! Gender mapping between FHIR administrative-gender codes and the openEHR
//! local value set.

pub fn gender_to_openehr(fhir_gender: Option<&str>) -> String {
    match fhir_gender.unwrap_or("").trim().to_lowercase().as_str() {
        "male" => "male",
        "female" => "female",
        "other" => "intersex",
        _ => "unknown",
    }
    .to_string()
}

/// Returns `None` when the input is empty (meaning "do not set gender").
pub fn gender_to_fhir(openehr_gender: Option<&str>) -> Option<String> {
    match openehr_gender.unwrap_or("").trim().to_lowercase().as_str() {
        "male" => Some("male".to_string()),
        "female" => Some("female".to_string()),
        "intersex" => Some("other".to_string()),
        "unknown" => Some("unknown".to_string()),
        "" => None,
        _ => Some("unknown".to_string()),
    }
}

File: go/bridge/gender.go
package bridge

import "strings"

// GenderToOpenEhr maps a FHIR administrative-gender code to the openEHR local code.
func GenderToOpenEhr(fhirGender string) string {
	switch strings.ToLower(strings.TrimSpace(fhirGender)) {
	case "male":
		return "male"
	case "female":
		return "female"
	case "other":
		return "intersex"
	default:
		return "unknown"
	}
}

// GenderToFhir maps an openEHR gender code to a FHIR administrative-gender code.
// An empty input returns "" (meaning "do not set gender").
func GenderToFhir(openEhrGender string) string {
	switch strings.ToLower(strings.TrimSpace(openEhrGender)) {
	case "male":
		return "male"
	case "female":
		return "female"
	case "intersex":
		return "other"
	case "unknown":
		return "unknown"
	case "":
		return ""
	default:
		return "unknown"
	}
}
